File: csharp/UpgradeBot.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace UpgradeBot
{
    class Planet
    {
        public int Id { get; set; }
        public int Owner { get; set; }
        public int GrowthRate { get; set; }
        public int NumShips { get; set; }
        public int X { get; set; }
        public int Y { get; set; }

    }

    internal class World
    {
        public int PlayerId { get; set; }
        public IList<Planet> Planets { get; set; }

        public World()
        {
            Planets = new List<Planet>();
        }

        public IEnumerable<Planet> PlanetsByOwner(int playerId)
        {
            return from planet in Planets
                   where planet.Owner == playerId
                   select planet;
        }
    }

    internal class UpgradeBot
    {
        private static IEnumerable<String> TakeTurns(World world)
        {
            var turns = new List<string>();

            var boost = (from p in world.PlanetsByOwner(world.PlayerId)
                         where (p.GrowthRate <= 6) && ((1 << p.GrowthRate) <= p.NumShips)
                         select p).FirstOrDefault();


            if (null != boost)
            {
                turns.Add(String.Format("B {0}", boost.Id));
                boost.NumShips -= (1 << boost.GrowthRate);
            }

            var force = (from source in world.PlanetsByOwner(world.PlayerId)
                         from target in world.PlanetsByOwner(0).OrderBy(x => x.NumShips)
                         where target.NumShips < source.NumShips
                         select new { source, target }).FirstOrDefault();

            if (null != force)
            {
                turns.Add(String.Format("F {0} {1} {2}", force.source.Id, force.target.Id, force.target.NumShips + 1));
                force.source.NumShips -= force.target.NumShips + 1;
            }
            return turns;
        }

        private static World ReadWorld()
        {
            var world = new World();
            var doRead = true;
            while (doRead)
            {
                var line = Console.In.ReadLine();
                if (null == line) break;
                var tokens = line.Trim().Split();
                switch (tokens.First())
                {
                    case "P":
                        {
                            var parameters = tokens.Skip(1).Select(int.Parse).ToArray();
                            var planet = new Planet
                                             {
                                                 Id = parameters[0],
                                                 X = parameters[1],
                                                 Y = parameters[2],
                                                 GrowthRate = parameters[3],
                                                 Owner = parameters[4],
                                                 NumShips = parameters[5]
                                             };
                            world.Planets.Add(planet);
                            break;
                        }

                    case "Y":
                        {
                            var parameters = tokens.Skip(1).Select(int.Parse).ToArray();
                            world.PlayerId = parameters[0];
                            break;
                        }

                    case ".":
                        doRead = false;
                        break;

                }
            }
            return world;
        }

        private static void WriteTurns(IEnumerable<String> turns)
        {
            foreach (var turn in turns)
            {
                Console.Out.WriteLine(turn);
            }
            Console.Out.WriteLine(".");
            Console.Out.Flush();
        }


        static void Main(string[] args)
        {
            while (true)
            {
                var world = ReadWorld();
                var turns = TakeTurns(world);
                WriteTurns(turns);
            }
        }
    }
}

File: ruby/upgrade_bot.rb
#!/usr/bin/env ruby

class Planet
  attr_reader :id, :growth_rate, :x, :y, :owner
  attr_accessor :num_ships

  def initialize(id, x, y, growth_rate, owner, num_ships)
    @id = id
    @x, @y = x, y
    @growth_rate = growth_rate
    @owner = owner
    @num_ships = num_ships
  end

  def remove_ships(n)
    @num_ships -= n
  end
end

class World
  attr_reader :planets
  attr_accessor :player_id

  def initialize
    @planets = []
  end

  def add_planet(planet)
    @planets << planet
  end

  def planets_by_owner(player_id)
    planets.select {|planet| planet.owner == player_id}
  end
end

def read_world()
  world = World.new
  while true
    case $stdin.readline.strip
    when "."
      break
    when /^P/
      planet_id, x, y, growth_rate, owner, num_ships = $_.split.drop(1).map(&:to_i)
      planet = Planet.new(planet_id, x, y, growth_rate, owner, num_ships)
      world.add_planet(planet)
    when /^Y/
      player_id, =  $_.split.drop(1).map(&:to_i)
      world.player_id = player_id
    end
  end
  world
end

def take_turns(world)
  turns = []

  boost_targets = world.planets_by_owner(world.player_id).select do |planet|
    (1 << planet.growth_rate) <= planet.num_ships && planet.growth_rate <= 6
  end

  if !boost_targets.empty?
    planet = boost_targets.first
    turns << "B #{planet.id}"
    planet.remove_ships(1 << planet.growth_rate)
  end

  victim = world.planets_by_owner(0).sort_by(&:num_ships).first
  if victim
    assault = world.planets_by_owner(world.player_id).select{|p| victim.num_ships < p.num_ships}.first
    if assault
      turns << "F #{assault.id} #{victim.id} #{victim.num_ships + 1}"
      assault.remove_ships(victim.num_ships + 1)
    end
  end

  turns
end

def write_turns(turns)
  turns.each { |t| $stdout.puts(t) }
  $stdout.puts "."
  $stdout.flush
end

begin
  while true
    world = read_world
    turns = take_turns(world)
    write_turns(turns)
  end
end
